Modify RealtimeGC for moving into OMR

jason-hall opened the pull request:
Signed-off-by: Jason Hall <jasonhal@ca.ibm.com>

jason-hall commented:
Do we want to rename all of the UDATA usage as well? Other OMR code kept them so I thought I should ask

dmitripivkine commented:
> > I believe we can keep old J9 types as UDATA for code in OpenJ9.
> 
> These files are getting changed in preparation for moving them directly into OMR

Types in files intended to be moved to OMR have to be changed. Types in code left in OpenJ9 (even in new files) might stay old. It makes easier to review (less changed lines).

dmitripivkine commented:
Error in Travis-ci build:
```
../../gc_staccato/StaccatoAccessBarrier.cpp: In member function ‘void MM_StaccatoAccessBarrier::scanContiguousArray(MM_EnvironmentRealtime*, J9IndexableObject*)’:
../../gc_staccato/StaccatoAccessBarrier.cpp:252:18: error: ‘class MM_RealtimeGC’ has no member named ‘isDynamicClassUnloadingEnabled’
  if(_realtimeGC->isDynamicClassUnloadingEnabled()) {
                  ^
make[6]: *** [gc_staccato/CMakeFiles/j9staccato.dir/StaccatoAccessBarrier.cpp.o] Error 1
gc_staccato/CMakeFiles/j9staccato.dir/build.make:88: recipe for target 'gc_staccato/CMakeFiles/j9staccato.dir/StaccatoAccessBarrier.cpp.o' failed
```

dmitripivkine commented on runtime/gc_glue_java/RealtimeGCDelegate.cpp:
```diff
@@ -0,0 +1,697 @@
+/*******************************************************************************
+ * Copyright (c) 2019 IBM Corp. and others
```
should be `Copyright (c) 2019, 2019 IBM Corp. and others`... here and for all new files

dmitripivkine commented:
Question for discussion: should `MM_RealtimeGCDelegate` be a child of `MM_GlobalCollectorDelegate`? Is there a reason to have _delegate and _realtimeDelegate both at the same time? I understand that there are not so much common on calls level (historically different implementation) but logically I believe it should be. It would create possibility to make code more universal in the future. For example class unloading for Realtime in `MM_RealtimeGCDelegate::processDyingClasses()` can use Standard/Balanced core code from `MM_GlobalCollectorDelegate::unloadDeadClassLoaders()`.

dmitripivkine commented:
Copyright in `runtime/gc_staccato/StaccatoAccessBarrier.cpp` is not updated

jason-hall commented:
> Question for discussion: should `MM_RealtimeGCDelegate` be a child of `MM_GlobalCollectorDelegate`?
I suppose this would mean changing the MM_GlobalCollectorDelegate in MM_GlobalCollector to a pointer, then casting it to a MM_RealtimeGCDelegate when using it?  Am I missing a simpler way?

dmitripivkine commented:
> > Question for discussion: should `MM_RealtimeGCDelegate` be a child of `MM_GlobalCollectorDelegate`?
> 
> I suppose this would mean changing the MM_GlobalCollectorDelegate in MM_GlobalCollector to a pointer, then casting it to a MM_RealtimeGCDelegate when using it? Am I missing a simpler way?

Ok, forget about it for now. Currently all three delegates (global, realtime and staccato) point to file where correspondent code lives. However there is no logical difference between all three - not like in Standard collectors. This issue might be addressed in future if necessary

dmitripivkine commented on runtime/gc_glue_java/RealtimeGCDelegate.hpp:
```diff
@@ -0,0 +1,124 @@
+#include "EnvironmentRealtime.hpp"
+#include "GCExtensions.hpp"
+
+class MM_RealtimeGCDelegate
```
Make sense to make this class be a child of class `MM_BaseNonVirtual` for better DDR support:
`class MM_RealtimeGCDelegate : public MM_BaseNonVirtual {`

dmitripivkine commented on runtime/gc_glue_java/StaccatoGCDelegate.hpp:
```diff
@@ -0,0 +1,64 @@
+
+class MM_StaccatoGC;
+
+class MM_StaccatoGCDelegate
```
Make sense to make this class be a child of class `MM_BaseNonVirtual` for better DDR support:
`class MM_StaccatoGCDelegate : public MM_BaseNonVirtual {`

dmitripivkine commented on runtime/gc_realtime/RealtimeGC.cpp:
```diff
@@ -278,11 +163,7 @@ void
-	if(NULL != _sched) {
-		_sched->kill(env);
-		_sched = NULL;
-	}
```
Is this part lost?

jason-hall replied:
I guess so. Good catch

dmitripivkine commented on runtime/gc_realtime/RealtimeGC.cpp:
```diff
@@ -393,8 +225,7 @@ MM_RealtimeGC::workerSetupForGC(MM_EnvironmentBase *env)
 void
 MM_RealtimeGC::clearGCStats(MM_EnvironmentBase *env)
 {
-	_extensions->globalGCStats.clear();
```
Why globalGCStats() is moved to delegate? Looks like it is called from OMR for Standard collectors.

jason-hall replied:
You're right, it can stay here instead of moving with java stats clear

dmitripivkine commented on runtime/gc_realtime/RealtimeGC.cpp:
```diff
@@ -975,17 +423,17 @@ MM_RealtimeGC::internalPostCollect(MM_EnvironmentBase *env, MM_MemorySubSpace *s
 	
 	/* Check if user overrode the default minimumFreeEntrySize */
-	if (_extensions->minimumFreeEntrySize != (UDATA)-1) {
+	if (_extensions->minimumFreeEntrySize != (uintptr_t)-1) {
```
you can use UDATA_MAX instead of (uintptr_t)-1

dmitripivkine commented on runtime/gc_realtime/RealtimeGC.cpp:
```diff
@@ -1036,13 +484,13 @@ MM_RealtimeGC::reportGCCycleFinalIncrementEnding(MM_EnvironmentBase *env)
+	uintptr_t approximateFreeFreeMemorySize;
+#if defined(OMR_GC_DYNAMIC_CLASS_UNLOADING)
 	MM_ClassUnloadStats *classUnloadStats = &_extensions->globalGCStats.classUnloadStats;
-#endif /* defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING) */
```
This is inconsistent with other code now. At the end of refactoring MM_RealtimeGC code should not have OMR_GC_DYNAMIC_CLASS_UNLOADING lines. However you can do this cleanup later.

dmitripivkine commented on runtime/gc_realtime/RealtimeGC.cpp:
```diff
@@ -1336,111 +730,72 @@ MM_RealtimeGC::condYield(MM_EnvironmentBase *env, U_64 timeSlackNanoSec)
 MM_RealtimeGC::isMarked(void *objectPtr)
 {
-	return _markingScheme->isMarked(static_cast<J9Object*>(objectPtr));
+	return _markingScheme->isMarked((omrobjectptr_t)(objectPtr));
```
Even in this form cast would not be necessary if input parameter type is changed from void* to omrobjectptr_t*
It can be done later in cleanup

dmitripivkine commented on runtime/gc_staccato/StaccatoGC.cpp:
```diff
@@ -231,127 +168,27 @@ MM_StaccatoGC::doTracing(MM_EnvironmentRealtime *env)
-					while (NULL != clazz) {
-						didWork |= _markingScheme->markClass(env, clazz);
-						clazz = _javaVM->internalVMFunctions->hashClassTableNextDo(&walkState);
+	return _staccatoDelegate.allocateAccessBarrier(env);
```
We should simplify call of allocateAccessBarrier() in future. There is no need to be so complicate

dmitripivkine commented on runtime/gc_staccato/StaccatoGC.cpp:
```diff
@@ -215,12 +156,8 @@ MM_StaccatoGC::doTracing(MM_EnvironmentRealtime *env)
-				_moreTracingRequired |= doClassTracing(env);
-			}
-#endif /* J9VM_GC_DYNAMIC_CLASS_UNLOADING */
+			_moreTracingRequired |= _staccatoDelegate.doTracing(env, &_realtimeDelegate);
```
I would add getRealtimeDelegate() to return &_realtimeDelegate

jason-hall commented:
I believe changes are up to date with latest comments

dmitripivkine commented:
Jenkins test sanity xLinux jdk8

dmitripivkine commented on runtime/gc_glue_java/RealtimeGCDelegate.hpp:
```diff
@@ -0,0 +1,125 @@
+	void reportClassUnloadingEnd(MM_EnvironmentBase *env);
+#endif /* J9VM_GC_DYNAMIC_CLASS_UNLOADING */
+
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
```
Next cleanup: closing "endif" and opening "if" can be removed to coalesce two sections together

dmitripivkine commented:
Next refactoring change should be collapsing MM_StaccatoGC to MM_RealtimeGC and  MM_StaccatoGCDelegate to MM_RealtimeGCDelegate with following clean up. There is no reason to keep them separate.

dmitripivkine commented:
Crash in cmdLineTester_jvmtitests_hcr_SE80_7:
```
16:49:16         Testing: rc018
16:49:16         Test start time: 2019/02/11 17:08:43 Atlantic Standard Time
16:49:16         Running command: "/home/jenkins/workspace/PullRequest-Sanity-JDK8-linux_x86-64_cmprssptrs-OpenJ9/build/linux-x86_64-normal-server-release/images/j2sdk-image/jre/bin/java" -Xcompressedrefs -Xgcpolicy:metronome -Xcompressedrefs  -Xdump    -agentlib:jvmtitest=test:rc018 -cp "/home/jenkins/workspace/PullRequest-Sanity-JDK8-linux_x86-64_cmprssptrs-OpenJ9/openj9/test/TestConfig/scripts/testKitGen/../../../../jvmtest/functional/cmdLineTests/jvmtitests/jvmtitest.jar:/home/jenkins/workspace/PullRequest-Sanity-JDK8-linux_x86-64_cmprssptrs-OpenJ9/openj9/test/TestConfig/scripts/testKitGen/../../../../jvmtest/TestConfig/lib/asm-all.jar" com.ibm.jvmti.tests.util.TestRunner
16:49:16         Time spent starting: 3 milliseconds
16:49:16         Time spent executing: 7566 milliseconds
16:49:16         Test result: FAILED
16:49:16         Output from test:
16:49:16          [OUT] *** Testing [1/1]:	testReflectRedefineAtSameTime
16:49:16          [OUT] starting reflect worker threads
16:49:16          [OUT] starting to populate java heap
16:49:16          [ERR] Unhandled exception
16:49:16          [ERR] Type=Segmentation error vmState=0x00020002
16:49:16          [ERR] J9Generic_Signal_Number=00000004 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
16:49:16          [ERR] Handler1=00007FA6FA506940 Handler2=00007FA6F98F1330 InaccessibleAddress=0000000000000018
16:49:16          [ERR] RDI=00007FA6A8003E40 RSI=0000000000000000 RAX=0000000000000001 RBX=00000000E0A30000
16:49:16          [ERR] RCX=00007FA6F20C2CF8 RDX=00007FA6F20C2D00 R8=0000000000000001 R9=00007FA6F403E130
16:49:16          [ERR] R10=000000000000000E R11=00000000E0800A98 R12=000000000000002E R13=0000000000000000
16:49:16          [ERR] R14=00007FA6A8001A68 R15=00007FA6F417C8C0
16:49:16          [ERR] RIP=00007FA6F82C8F41 GS=0000 FS=0000 RSP=00007FA6EF749A00
16:49:16          [ERR] EFlags=0000000000010246 CS=0033 RBP=00000000E0A30000 ERR=0000000000000004
16:49:16          [ERR] TRAPNO=000000000000000E OLDMASK=0000000000000000 CR2=0000000000000018
```

dmitripivkine commented:
Crash occur an attempt to scan class `!j9class 0x142f900` but it's class object has been collected:
```
> !j9class 0x142f900
J9Class at 0x142f900 {
  Fields for J9Class:
	0x0: UDATA eyecatcher = 0x0000000099669966 (2573637990)
	0x8: struct J9ROMClass* romClass = !j9romclass 0x00007FA6F45CBD60
	0x10: void** superclasses = !j9x 0x000000000142FBD8
	0x18: UDATA classDepthAndFlags = 0x00000000000E0003 (917507)
	0x20: U32 classDepthWithFlags = 0x00000000 (0)
	0x24: U32 classFlags = 0x00000000 (0)
	0x28: struct J9ClassLoader* classLoader = !j9classloader 0x00007FA6F42659D8
	0x30: struct J9Object* classObject = !j9object 0x00000000E0A30000 // <FAULT> <---- collected class object
	0x38: volatile UDATA initializeStatus = 0x0000000000000001 (1)
	0x40: struct J9Method* ramMethods = !j9method 0x000000000142FB98 // sun/reflect/GeneratedMethodAccessor5.<init>()V
	0x48: UDATA* ramStatics = !j9x 0x0000000000000000
	0x50: struct J9Class* arrayClass = !j9class 0x0000000000000000
	0x58: UDATA totalInstanceSize = 0x0000000000000004 (4)
	0x60: struct J9ITable* lastITable = !j9itable 0x000000000132AB10
	0x68: UDATA* instanceDescription = !j9x 0x0000000000000001
	0x70: UDATA* instanceLeafDescription = !j9x 0x0000000000000001
	0x78: UDATA instanceHotFieldDescription = 0x0000000000000000 (0)
	0x80: struct J9Method* initializerCache = !j9method 0x000000000142FB98 // sun/reflect/GeneratedMethodAccessor5.<init>()V
	0x88: UDATA romableAotITable = 0x00007FA6F3B968B0 (140355030313136)
	0x90: UDATA packageID = 0x00007FA6F45CBD61 (140355041017185)
	0x98: struct J9Module* module = !j9module 0x0000000000000000
	0xa0: struct J9Class* subclassTraversalLink = !j9class 0x000000000142F500 // sun/reflect/GeneratedMethodAccessor1
	0xa8: struct J9Class* subclassTraversalReverseLink = !j9class 0x000000000142FE00 // sun/reflect/GeneratedMethodAccessor2
	0xb0: void** iTable = !j9x 0x000000000132AB10
	0xb8: UDATA castClassCache = 0x000000000132E400 (20112384)
	0xc0: void** jniIDs = !j9x 0x0000000000000000
	0xc8: UDATA lockOffset = 0x0000000000000004 (4)
	0xd0: UDATA newInstanceCount = 0x0000000000000BB8 (3000)
	0xd8: IDATA backfillOffset = 0x0000000000000008 (8)
	0xe0: struct J9Class* replacedClass = !j9class 0x0000000000000000
	0xe8: UDATA finalizeLinkOffset = 0x0000000000000000 (0)
	0xf0: struct J9Class* nextClassInSegment = !j9class 0x0000000000000000
	0xf8: UDATA* ramConstantPool = !j9x 0x000000000142FA90
	0x100: struct J9Object** callSites = !j9x 0x0000000000000000
	0x108: struct J9Object** methodTypes = !j9x 0x0000000000000000
	0x110: struct J9Object** varHandleMethodTypes = !j9x 0x0000000000000000
	0x118: struct J9VMCustomSpinOptions* customSpinOption = !j9vmcustomspinoptions 0x0000000000000000
	0x120: void** staticSplitMethodTable = !j9x 0x0000000000000000
	0x128: void** specialSplitMethodTable = !j9x 0x0000000000000000
	0x130: struct J9JITExceptionTable* jitMetaDataList = !j9jitexceptiontable 0x0000000000000000
	0x138: struct J9Class* gcLink = !j9class 0x0000000000000000
	0x140: struct J9Class* hostClass = !j9class 0x000000000142F900 // sun/reflect/GeneratedMethodAccessor5
}
Class name: sun/reflect/GeneratedMethodAccessor5
To view static fields, use !j9statics 0x000000000142F900
To view instance shape, use !j9classshape 0x000000000142F900

> !j9x 0x00000000E0A30000
0xE0A30000 :  0000000000000001 0000000000000058 [ ........X....... ]
0xE0A30010 :  00000000e0042760 0000000000000000 [ `'.............. ]
```

dmitripivkine commented:
Jenkins test sanity xLinux jdk8

dmitripivkine commented:
Jenkins test extended aix jdk11

dmitripivkine commented:
Jenkins test compile win jdk11

dmitripivkine commented:
Jenkins compile win jdk11

dmitripivkine commented:
Jenkins test sanity xLinux jdk11

dmitripivkine commented:
Jenkins test sanity xLinux jdk8

dmitripivkine commented:
rebase to latest does not help: `rc018` in `cmdLineTester_jvmtitests_hcr_7` still be broken

dmitripivkine commented:
Jenkins test sanity xLinux jdk8

dmitripivkine commented:
Jenkins test sanity aix jdk11

dmitripivkine commented:
Jenkins test sanity xLinux jdk8

dmitripivkine commented:
Jenkins test sanity xLinux jdk8

dmitripivkine commented:
now failed in:
```
16:12:39         Testing: rc018
16:12:39         Test start time: 2019/02/19 14:47:23 Central Standard Time
16:12:39         Running command: "/home/u0020236/workspace/PullRequest-Sanity-JDK11-aix_ppc-64_cmprssptrs-OpenJ9/build/aix-ppc64-normal-server-release/images/jdk/bin/java" -Xcompressedrefs -Xgcpolicy:metronome -Xcompressedrefs  -Xdump    -agentlib:jvmtitest=test:rc018 -cp "/home/u0020236/workspace/PullRequest-Sanity-JDK11-aix_ppc-64_cmprssptrs-OpenJ9/openj9/test/TestConfig/scripts/testKitGen/../../../../jvmtest/functional/cmdLineTests/jvmtitests/jvmtitest.jar:/home/u0020236/workspace/PullRequest-Sanity-JDK11-aix_ppc-64_cmprssptrs-OpenJ9/openj9/test/TestConfig/scripts/testKitGen/../../../../jvmtest/TestConfig/lib/asm-all.jar" com.ibm.jvmti.tests.util.TestRunner
16:12:39         Time spent starting: 8 milliseconds
16:12:39         Time spent executing: 15772 milliseconds
16:12:39         Test result: FAILED
16:12:39         Output from test:
16:12:39          [OUT] *** Testing [1/1]:	testReflectRedefineAtSameTime
16:12:39          [OUT] starting reflect worker threads
16:12:39          [OUT] starting to populate java heap
16:12:39          [ERR] 20:47:25.226 0x3012fa00    j9mm.141    *   ** ASSERTION FAILED ** at RealtimeMarkingScheme.cpp:703: ((false))
16:12:39          [ERR] JVMDUMP039I Processing dump event "traceassert", detail "" at 2019/02/19 14:47:25 - please wait.
16:12:39          [ERR] JVMDUMP032I JVM requested System dump using '/home/u0020236/workspace/PullRequest-Sanity-JDK11-aix_ppc-64_cmprssptrs-OpenJ9/openj9/test/TestConfig/test_output_15506060312695/cmdLineTester_jvmtitests_hcr_7/core.20190219.144725.12583044.0001.dmp' in response to an event
```

This assertion means object type from j9class is unknown. Most likely treated garbage as an object.

dmitripivkine commented:
Failure with assertion in at RealtimeMarkingScheme.cpp:703 seems the same - number of class and ClassLoader objects are collected:

```
Checking CLASS HEAP...
  <gc check (1): from debugger: CLASS HEAP: slots slot 169f700(169f730) -> e1010160: corrupt data exception>
  <gc check (2): from debugger: CLASS HEAP: slots slot 169ee00(169ee30) -> e0c00160: object not in an object region>
  <gc check (3): from debugger: CLASS HEAP: slots slot 169e400(169e430) -> e09803c8: class pointer not in a class segment>

Checking CLASS LOADERS...
  <gc check (4): from debugger: CLASS LOADERS: slot 7f2a602658f0(7f2a60265908) -> e08b01b0: class pointer not in a class segment>
  <gc check (5): from debugger: CLASS LOADERS: slot 7f2a602659d8(7f2a602659f0) -> e0f00048: corrupt data exception>
  <gc check (6): from debugger: CLASS LOADERS: slot 7f2a60265c90(7f2a60265ca8) -> e1070048: object not in an object region>
```

dmitripivkine commented on runtime/gc_glue_java/StaccatoGCDelegate.hpp:
```diff
@@ -0,0 +1,65 @@
+class MM_StaccatoGCDelegate : public MM_BaseNonVirtual
+{
+private:
+	OMR_VM *_vm;
```
Do we really need to have (OMR_VM *) in delegates? Looks like (J9JavaVM*) would be better choice. Did I missed something?

dmitripivkine commented on runtime/gc_realtime/RealtimeGC.cpp:
```diff
@@ -806,51 +303,8 @@ MM_RealtimeGC::incrementalCollect(MM_EnvironmentRealtime *env)
-			|| fixupForClassUnload) {
-		_fixHeapForWalk = true;
-	}	
+	_realtimeDelegate.incrementalCollect(env);
```
`_dynamicClassUnloadingEnabled ` use to be set before Mark. Currently however it has been moved to Delegate and set after Mark

jason-hall commented:
- Returned setting of _dynamicClassUnloadingEnabled in incrementalCollect() back to original order
- Returned initialization of extensions->accessBarrier back to original code
- Removed unnecessary OMR_VM pointers from MM_RealtimeGCDelegate and MM_StaccatoDelegate
- Added missing omrcfg.h include for build flags in MM_RealtimeGC

dmitripivkine commented:
Jenkins test sanity xLinux jdk8

dmitripivkine commented:
Jenkins test sanity AIX jdk11

dmitripivkine commented:
Jenkins compile win

dmitripivkine commented:
Warnings treated as errors on Windows:
```
19:27:04  c:\users\jenkins\workspace\pullrequest-compile-jdk8-win_x86-64_cmprssptrs-openj9\build\windows-x86_64-normal-server-release\vm\gc_realtime\RealtimeGC.hpp(262) : error C2220: warning treated as error - no 'object' file generated
19:27:04  c:\users\jenkins\workspace\pullrequest-compile-jdk8-win_x86-64_cmprssptrs-openj9\build\windows-x86_64-normal-server-release\vm\gc_realtime\RealtimeGC.hpp(262) : warning C4355: 'this' : used in base member initializer list
19:27:05  Packet.cpp
19:27:05  C:\Users\jenkins\workspace\PullRequest-Compile-JDK8-win_x86-64_cmprssptrs-OpenJ9\build\windows-x86_64-normal-server-release\vm\gc_staccato\StaccatoGC.hpp(57) : warning C4355: 'this' : used in base member initializer list
```

dmitripivkine commented:
Jenkins test sanity xLinux jdk8

dmitripivkine commented:
Jenkins compile win

jason-hall commented:
> stderr: error: RPC failed; curl 18 transfer closed with outstanding read data remaining
> fatal: The remote end hung up unexpectedly

dmitripivkine commented:
Jenkins test sanity xLinux jdk8

dmitripivkine commented:
Jenkins test sanity AIX jdk11

dmitripivkine commented:
Jenkins test sanity xLinux jdk8

dmitripivkine commented:
Jenkins compile win

dmitripivkine commented:
Jenkins test sanity xLinux jdk8

jason-hall commented:
I don't see how this failure could be related

dmitripivkine commented:
Jenkins compile win

dmitripivkine commented:
Jenkins test sanity xLinux jdk8

dmitripivkine commented:
Jenkins test sanity AIX jdk11

charliegracie commented:
I will complete my review on Monday March 4th
